README.md
This package makes creating drag and drop UI easier. If you are planning to make an app where you need to drag & drop the matching component, this package would be helpful for you.

## Features

Create drag and drop UI's with few lines of code
Useful for games and other interactive apps

![](./screenshot.gif)

## Methods

DragDropItem
  This is the parent class. It accepts various parameters.

key:
value:
dragChild:
dropChild:
iconData:

Except for key and value , all fields are optional. If no dragChild/dropChild is provided, you'll get a Text widget as your drag/drop components.

DragDropWidget is the base class where you provide reference to your drag drop items. It accepts 3 parameters.
items: Collection of DragDropItem.
onMatched: If the dragged item matched with dropped item.
onMismatched : if the dragged item is not matched with drop items.


## Getting started

Add the packages to pubspec and get started.

drag_drop_match_widget: ^0.0.1

## Usage

```dart

final items = [
  DragDropItem(
      key: "apple",
      value: "apple",
      dragChild: Text(
        "APPLE",
        style: TextStyle(fontSize: 30),
      ),
      dropChild: Text(
        "🍎",
        style: TextStyle(fontSize: 50),
      ),
      iconData: Icons.one_k),
  DragDropItem(
      key: "ball",
      value: "Ball",
      dragChild: Text(
        "BALL",
        style: TextStyle(fontSize: 30),
      ),
      dropChild: Text(
        "⚽",
        style: TextStyle(fontSize: 50),
      ),
      iconData: Icons.one_k),
  DragDropItem(
      key: "cat",
      value: "Cat",
      dragChild: Text(
        "CAT",
        style: TextStyle(fontSize: 30),
      ),
      dropChild: Text(
        "😺",
        style: TextStyle(fontSize: 50),
      ),
      iconData: Icons.one_k)
];
```
Use the widget as,

```
DragDropWidget(
              items: items,
              //on  Matched event
              onMatched: (DragDropItem? item) {
                score += 10;
                setState(() {});
                matched.add(item);
                
                if (matched.length == items.length) {
                  print("GAME OVER");
                }
              },
              //on mismatched event
              onMisMatched: (DragDropItem? item) {
                print("Mismatched");
                print(item);
              });

```

## Additional information

To find more information about the package visit https://github.com/sabine33/flutter_drag_drop_match_widget


CHANGELOG.md
## 0.0.1

* Initial Release

test/drag_drop_match_widget_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:drag_drop_match_widget/drag_drop_match_widget.dart';

void main() {
  test('adds one to input values', () {});
}


lib/drag_drop_model.dart
import 'package:flutter/material.dart';

class DragDropItem {
  final String key;
  final String value;
  Widget? dragChild;
  Widget? dropChild;
  IconData? iconData;
  Widget? feedbackItem;
  Widget? childWhenDragging;
  bool willAccept = true;
  bool isAccepted;
  DragDropItem(
      {required this.key,
      required this.value,
      this.iconData,
      this.dragChild,
      this.dropChild,
      this.feedbackItem,
      this.childWhenDragging,
      this.isAccepted = false}) {
    feedbackItem =
        this.dragChild ?? Icon(iconData, size: 30, color: Colors.teal);
    dragChild = this.dragChild ??
        Text(key,
            style: TextStyle(
              fontSize: 20,
            ));
    dropChild = dropChild ??
        Text(value,
            style: TextStyle(
              fontSize: 20,
            ));
    childWhenDragging = dragChild ??
        Text(value,
            style: TextStyle(
              fontSize: 20,
            ));
  }

  factory DragDropItem.fromMap(MapEntry map) {
    return DragDropItem(
        key: map.key,
        value: map.value,
        dragChild: Text(map.key),
        dropChild: Text(map.value));
  }
}


lib/drag_drop_match_widget.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

class DragDropItem {
  //unique key for matching source and target
  //Accepts string
  final String key;
  //Value of a matching target
  //Accepts string
  final String value;
  //Widget to put into draggable
  //Any valid widget, else Text widget is used
  Widget? dragChild;
  //Widget to put into drop field/target
  //Any valid widget, else Text widget is used
  Widget? dropChild;
  //Icon for widget , [Optional]
  IconData? iconData;
  //Feedback widget on dragging, at the pointer
  Widget? feedbackItem;
  //Child to leave at source when dragging
  Widget? childWhenDragging;
  //whether a drop target will accept this model or not
  //[boolean]
  bool willAccept = true;
  //whether the model is accepted or not
  //locked if isAccepted is true, else not
  bool isAccepted;
  //default text style for a drag/drop children
  TextStyle? defaultTextStyle;

  DragDropItem(
      {required this.key,
      required this.value,
      this.iconData,
      this.dragChild,
      this.dropChild,
      this.feedbackItem,
      this.childWhenDragging,
      this.isAccepted = false,
      this.defaultTextStyle}) {
    feedbackItem = Material(
        color: Colors.transparent,
        child: dragChild ?? Icon(iconData, size: 30, color: Colors.teal));
    dragChild = dragChild ??
        Text(key, style: TextStyle(fontSize: 20, color: Colors.red));
    dropChild = dropChild ??
        Text(value,
            style: TextStyle(
              fontSize: 20,
            ));
    childWhenDragging = dragChild ??
        Text(value,
            style: TextStyle(
              fontSize: 20,
            ));
    this.defaultTextStyle = _defaultTextStyle;
  }

  factory DragDropItem.fromMap(MapEntry map) {
    return DragDropItem(
        key: map.key,
        value: map.value,
        dragChild: Text(map.key, style: _defaultTextStyle),
        dropChild: Text(map.value, style: _defaultTextStyle));
  }
}

//dragdrop action , matched or not matched
typedef DragDropAction = void Function(DragDropItem item);
//drag drop default child style
final _defaultTextStyle = TextStyle(fontSize: 50);
final _matchedWidget = Text("✅", style: _defaultTextStyle);

///Drag drop widget for creating match views
///Input your stuffs on [items]
///[onMatched] is handler for matched event
///[onMisMatched] is handler for not matched event
class DragDropWidget extends StatefulWidget {
  //list of items you want to embed in drag drop UI
  final List<DragDropItem> items;
  //on matched event
  final DragDropAction onMatched;
  //on mismatched event
  final DragDropAction onMisMatched;
  DragDropWidget(
      {Key? key,
      required this.items,
      required this.onMatched,
      required this.onMisMatched})
      : super(key: key);

  @override
  State<DragDropWidget> createState() => _DragDropWidgetState();
}

class _DragDropWidgetState extends State<DragDropWidget> {
  //The backdrop style : after drag drop is accepted
  bool isLoaded = false;

  @override
  void initState() {
    super.initState();
  }

  List<DragDropItem> clonedList() {
    var newList = widget.items.toList();
    newList.shuffle();

    return newList;
  }

  @override
  Widget build(BuildContext context) {
    // dropLists = dropList();

    return Material(
        child: Row(
      // mainAxisAlignment: MainAxisAlignment.spaceBetween,
      children: [
        Column(
            children: widget.items.map((DragDropItem item) {
          return Container(
            margin: const EdgeInsets.all(8.0),
            child: Draggable<DragDropItem>(
                data: item,
                childWhenDragging: item.dragChild!,
                feedback: item.feedbackItem!,
                child:
                    item.isAccepted == true ? _matchedWidget : item.dragChild!),
          );
        }).toList()),
        Spacer(),
        Column(
            children: clonedList().map((DragDropItem item) {
          return DragTarget<DragDropItem>(
            onAccept: (receivedItem) {
              if (receivedItem.key == item.key) {
                setState(() {
                  item.isAccepted = true;
                });

                print("ACCEPTED");
                widget.onMatched(receivedItem);
              }
            },
            onLeave: (receivedItem) {
              print("NOT ACCEPTED");
              item.willAccept = false;

              widget.onMisMatched(receivedItem!);
            },
            onWillAccept: (receivedItem) {
              bool willAccept =
                  receivedItem?.key == item.key && !item.isAccepted == true;
              item.willAccept = willAccept;

              return willAccept;
            },
            builder: (context, acceptedItems, rejectedItem) => Container(
                alignment: Alignment.center,
                margin: const EdgeInsets.all(8.0),
                child: Container(
                    color: item.isAccepted ? Colors.green : Colors.transparent,
                    child: item.dropChild)),
          );
        }).toList())
      ],
    ));
  }
}


example/lib/main.dart
import 'package:drag_drop_match_widget/drag_drop_match_widget.dart';
import 'package:flutter/material.dart';

var map = {
  'APPLE': '🍎',
  'BALL': '⚽',
  'CAT': '😺',
  'DOG': '🐶',
  'ELEPHANT': '🐘',
  'FOX': '🦊',
  'FISH': '🐟',
};

var items_2 = [
  DragDropItem(
      key: "apple",
      value: "apple",
      dragChild: Text(
        "APPLE",
        style: TextStyle(fontSize: 30),
      ),
      dropChild: Text(
        "🍎",
        style: TextStyle(fontSize: 50),
      ),
      iconData: Icons.one_k),
  DragDropItem(
      key: "ball",
      value: "Ball",
      dragChild: Text(
        "BALL",
        style: TextStyle(fontSize: 30),
      ),
      dropChild: Text(
        "⚽",
        style: TextStyle(fontSize: 50),
      ),
      iconData: Icons.one_k),
  DragDropItem(
      key: "cat",
      value: "Cat",
      dragChild: Text(
        "CAT",
        style: TextStyle(fontSize: 30),
      ),
      dropChild: Text(
        "😺",
        style: TextStyle(fontSize: 50),
      ),
      iconData: Icons.one_k)
];
void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  List<DragDropItem> items = [];
  int score = 0;
  var matched = [];

  @override
  void initState() {
    items_2.shuffle();
    // map.entries.forEach((element) {
    //   items.add(DragDropItem.fromMap(element)
    //     ..defaultTextStyle = TextStyle(fontSize: 20, color: Colors.red));
    // });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Matching Game',
        debugShowCheckedModeBanner: false,
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: Scaffold(
          appBar: AppBar(
            title: Text("Score $score"),
          ),
          body: Container(
            child: DragDropWidget(
                items: items_2,
                onMatched: (DragDropItem? item) {
                  score += 10;
                  setState(() {});
                  matched.add(item);
                  if (matched.length == items.length) {
                    print("GAME OVER");
                  }
                },
                onMisMatched: (DragDropItem? item) {
                  print("Mismatched");
                  print(item);
                }),
          ),
        ));
  }
}
